### ISSUE TITLE
Permit type definition per scope

### ISSUE THREAD

**Issue by kingori** [InsertKoinIO/koin#412]
**Is your feature request related to a problem? Please describe.**
I plan to migrate dagger2 to koin. In dagger, I can declare type definition per scope. But in Koin, DefinitionOverrideException is thrown. I think this feature is useful for some cases.

```kotlin
val koin = koinApplication {
            modules(
                module {
                    scope(named("a")) {
                        scoped { ModelA("a") }
                    }

                    //DefinitionOverrideException thrown here
                    scope(named("b")) {
                        scoped { ModelA("b") }
                    }
                }
            )
}

val scopeA = koin.koin.getOrCreateScope("id_a", named("a"))
val scopeB = koin.koin.getOrCreateScope("id_b", named("b"))
     
koin.koin.get<ModelA>(scope = scopeA).let {
  Assert.assertEquals("A", it.name)
}

koin.koin.get<ModelA>(scope = scopeB).let {
  Assert.assertEquals("B", it.name)
}
```

I read your reponse at stackoverflow. You suggested to raise an issue on it. So I did. :-)
https://stackoverflow.com/a/52225309/676215

**Describe the solution you'd like**
Scoped bean definition is only visible inside scope. So, definition of same type can be exsited on multiple scopes.

**Describe alternatives you've considered**
Define different type for each scope like below.
```kotlin
               module {
                    scope(named("a")) {
                        scoped { ModelAForScopeA("a") }
                    }

                    //DefinitionOverrideException thrown here
                    scope(named("b")) {
                        scoped { ModelAForScopeB("b") }
                    }
                }
```

**Target Koin project**
2.0.0 ?

**Comment by arnaudgiuliani** [comment]
your proposal is about having same type definition, but from different scope? to avoid having to use naming/qualifier

**Comment by kingori** [comment]
@arnaudgiuliani Yes, right.

**Comment by arnaudgiuliani** [comment]
is available in current 2.0.1
